### ISSUE TITLE
[sdk/go] Make room for improved typing without breaking change

### ISSUE THREAD

**Issue by lukehoban** [pulumi/pulumi#4373]
Today, our `ApplyT` and `ApplyString` APIs are very loosely typed, and depend on runtime checks (with panics) for "type-checking":

```
func (o *OutputState) ApplyT(applier interface{}) Output
func (o *OutputState) ApplyString(applier interface{}) StringOutput
```

A key reason these are weakly typed is that they are implemented once on `OutputState` and "inherited" by all concrete `Output` structs via struct embedding.  As a result, the argument type of the callback can't be known statically, and so the feeling was it was better to allow users to write the natural `func (s string) (string, error)` and eagerly runtime-check it instead of forcing `func (s interface{}) (interface{}, error)` (which is still available as `.Apply` everywhere).

We can though get the best of both worlds if we "override" the implementation of `ApplyT`/`ApplyString` on at least all primitives.

That is - in addition to the implementations above (don't remove anything), also add:

```
func (o *String) ApplyT(f func(s string) (interface{}, error)) Output
func (o *String) ApplyString(f func(s string) (string, error)) StringOutput
```

This will then accept the same things that are accepted today - but will strongly type check them (as well as give much better hints to users about what's expected).

There is one small hole in this though - today the weakly-typed API accepts two forms of this:

```
// The applier function must have one of the following signatures:
//
//    func (v U) T
//    func (v U) (T, error)
```

There is no way to support both of these forms in a strongly typed Go API.  

A suggestion is to remove the first of the forms above from the `2.0` release, so that we have the option to provide strongly typed `Apply*` overloads in the future.

**Comment by pgavlin** [comment]
> A suggestion is to remove the first of the forms above from the 2.0 release, so that we have the option to provide strongly typed Apply* overloads in the future.

I think removing this form would be a _huge_ usability hit, to be honest. I don't think I've written more than a handful of `ApplyT` calls that use the second form.

**Comment by lukehoban** [comment]
> I think removing this form would be a huge usability hit, to be honest. I don't think I've written more than a handful of ApplyT calls that use the second form.

I hear you - but in practice this is *extremely* common for Go API design.  Had we not ever had this loosely typed implementation, I don't think we would have blinked twice at having a required `error` in the return type.

That said, it does mean that this is definitely a non-trivial breaking change we'd have to introduce ahead of `2.0` (though a pretty mechanical change to make).

But the value of having a "normal-looking" `Apply*` API in the future seems very significant.

**Comment by pgavlin** [comment]
> But the value of having a "normal-looking" Apply* API in the future seems very significant.

Are you proposing that we would generate these methods for every Output type, or just for builtins?

**Comment by lukehoban** [comment]
> Are you proposing that we would generate these methods for every Output type, or just for builtins?

For builtins.  And then as part of "API design guidelines" note that new `Output` types can optionally choose to implement these for stronger typing.

**Comment by lukehoban** [comment]
FWIW - An alternative to all this would be to add `ApplyStringString`/`ApplyStringT`/`ApplyStringBool` with the strong typing (first type the input type, second type the output type).  I prefer the recommendation in this issue over that - as it just makes the API surface area we already have more strongly typed.

**Comment by joeduffy** [comment]
I agree very much that we should add strongly typed versions here.

In principle, I see no reason not to support different overloads for errors vs non-errors:

```go
func (o *String) ApplyT(f func(s string) interface{}) Output
func (o *String) ApplyTErr(f func(s string) (interface{}, error)) Output
func (o *String) ApplyString(f func(s string) string) StringOutput
func (o *String) ApplyStringErr(f func(s string) (string, error)) StringOutput
```

Note, however, that I don't quite understand how the error paths work, given that I don't believe outputs in our other runtimes support the notion of "error" (unlike promises, outputs can't be "broken").

**Comment by pgavlin** [comment]
> (unlike promises, outputs can't be "broken").

This is not correct. Outputs can be broken if they are resolved by callbacks that fail.

**Comment by pgavlin** [comment]
FWIW, the primary improvement I see in adding these methods is in a better IDE experience. We could implement static checking for `Apply` via a linter (ala the support for checking format strings in `govet`). I think that adding all of these methods is going to make the raw documentation increasingly unusable, though: the more of these methods we add, the noisier it gets. We also lose some uniformity, as not all outputs will have all of these methods unless we code-generate them in the SDKs.

**Comment by lukehoban** [comment]
> FWIW, the primary improvement I see in adding these methods is in a better IDE experience. We could implement static checking for Apply via a linter (ala the support for checking format strings in govet).

I mean - it's just plain normal value of static typing.  Saying that static typing is only useful for "a better IDE experience" doesn't feel quite right to me - though even if it was - I think that's a non-trivial part of the value of Pulumi.  We aren't going to see a large % of Pulumi users pick up a random linter we add - they expect building their code to catch static type errors.

**Comment by lukehoban** [comment]
> We also lose some uniformity, as not all outputs will have all of these methods unless we code-generate them in the SDKs.

FWIW - A key thing to my proposal (different from @joeduffy's) - is that we are not adding any new methods at all.  We are just allowing types to opt in to stronger typing.  This does not actually expand the number of user-visible methods.

**Comment by EvanBoyle** [comment]
I really like this idea. We can generate additional methods like this: 

```go
// ApplyInt is like ApplyT, but returns a IntOutput.
func (o *StringOutput) ApplyInt(applier func(v string) int) IntOutput {
	return o.ApplyInt(applier)
}
```

I tested this locally just to make sure I could get autocomplete to work properly for a `String` specific ApplyString. It looks like in order for autocomplete in VS code to pick this up me need a specific implementation for each form (String, StringOutput, StringPtrOutput, ...):

```go
// ApplyString is like ApplyT, but returns a StringOutput.
func (o *StringOutput) ApplyString(applier func(v string) string) StringOutput {
	return o.ApplyString(applier)
}
```

![Screen Shot 2020-04-13 at 12 50 19 PM](https://user-images.githubusercontent.com/2553171/79155494-0b7c3300-7d86-11ea-9baf-d59edc2fdf83.png)

**Comment by EvanBoyle** [comment]
We could be emitting a huge number of new methods here. This might further impact compile/link performance?

**Comment by EvanBoyle** [comment]
Also note that the old forms are inaccessible unless you specify `o2.OutputState`:

![Screen Shot 2020-04-13 at 1 00 03 PM](https://user-images.githubusercontent.com/2553171/79155946-cc9aad00-7d86-11ea-8b0e-c73c105833fe.png)

To use the old forms you'll have to specify:

```go
o2.OutputState.ApplyString(func(v interface{}) interface{} { return "" })
```

So this definitely will be a breaking change no matter which decision we make here, right?

**Comment by EvanBoyle** [comment]
If we're going to further impact compile/link/startup times, I don't think splitting out the error vs non-error forms of this function provides much value to the user in exchange for the cost of significantly increasing the number of generated methods. Gophers are used to checking error values. I don't see a problem with making (interface{}, error) the only form that we accept.

**Comment by lukehoban** [comment]
> We could be emitting a huge number of new methods here. This might further impact compile/link performance?

It definitely could - I'm not sure.  It is indeed a lot more methods. 

> To use the old forms you'll have to specify

There is no benefit to trying to use "the old forms".  Anything you could have written without panicing would also be supported by the new form.  It's purely just more strongly staticly typed.  It is not otherwise observably different.  (Other than the point about whether to require an `error` output - which is why the proposal here focuses on that).

> So this definitely will be a breaking change no matter which decision we make here, right?

With the proposal I'm making, the only necessary breaking change would be to require an `error` output.  If we make that change prior to `2.0`, the rest of this change is a non-breaking addition we can make post-2.0.

We could choose other options here (like what @joeduffy suggests) which would be breaking or else more additive changes.

> I don't think splitting out the error vs non-error forms of this function provides much value to the user in exchange for the cost of significantly increasing the number of generated methods. Gophers are used to checking error values. I don't see a problem with making (interface{}, error) the only form that we accept.

I personally agree with this.

**Comment by EvanBoyle** [comment]
@lukehoban I'm not trying to say that there is a benefit to the old forms. Just that they are no longer accessible. For instance, you could write some non-specific code that uses `interface{}` and it compiles today:

```go
o2 := out.ApplyString(func(v int) string { return *new(string) })
_, known, _, err := await(o2)
assert.True(t, known)
assert.NoError(t, err)

o2 = o2.ApplyString(func(v interface{}) string { return *new(string) })
_, known, _, err = await(o2)
assert.True(t, known)
assert.NoError(t, err)
```

Anyone doing this today would experience the error that I showed above in the screenshot. I'll show the output again here:

```sh
Evans-MBP:pulumi evanboyle$ go test -run TestOutputApply
# github.com/pulumi/pulumi/sdk/go/pulumi [github.com/pulumi/pulumi/sdk/go/pulumi.test]
./alias.go:85:41: cannot call pointer method on parent.ToStringOutput()
./alias.go:85:41: cannot take the address of parent.ToStringOutput()
./types_builtins_test.go:1016:24: cannot use func literal (type func(interface {}) string) as type func(string) string in argument to o2.ApplyString
FAIL	github.com/pulumi/pulumi/sdk/go/pulumi [build failed]
Evans-MBP:pulumi evanboyle$ 
```

To make this code compile you now have to write:

```go
o2 := out.ApplyString(func(v int) string { return *new(string) })
_, known, _, err := await(o2)
assert.True(t, known)
assert.NoError(t, err)

o2 = o2.OutputState.ApplyString(func(v interface{}) string { return *new(string) })
_, known, _, err = await(o2)
assert.True(t, known)
assert.NoError(t, err)
```

**Comment by EvanBoyle** [comment]
Whether or not we want people doing this, they could be today and thus this is a break right?

**Comment by lukehoban** [comment]
> Whether or not we want people doing this, they could be today and thus this is a break right?

> o2 = o2.ApplyString(func(v interface{}) string { return *new(string) })

Good point - this one would no longer work as well - so you are right this is a (somewhat more minor) breaking change.
